action/protocol/staking/handler_candidate_endorsement.go

```diff
@@ -0,0 +1,103 @@
+package staking
+
+import (
+	"context"
+
+	"github.com/iotexproject/iotex-address/address"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	handleCandidateEndorsement = "candidateEndorsement"
+)
+
+func (p *Protocol) handleCandidateEndorsement(ctx context.Context, act *action.CandidateEndorsement, csm CandidateStateManager) (*receiptLog, []*action.TransactionLog, error) {
+	actCtx := protocol.MustGetActionCtx(ctx)
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	log := newReceiptLog(p.addr.String(), handleCandidateEndorsement, featureCtx.NewStakingReceiptFormat)
+
+	esm := NewEndorsementStateManager(csm.SM())
+	bucket, rErr := p.fetchBucket(csm, act.BucketIndex())
+	if rErr != nil {
+		return log, nil, rErr
+	}
+	cand := csm.GetByOwner(bucket.Candidate)
+	if cand == nil {
+		return log, nil, errCandNotExist
+	}
+	log.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), bucket.Candidate.Bytes(), []byte{byteutil.BoolToByte(act.Endorse())})
+
+	var err error
```

Comment:
move `esm := NewEndorsementStateManager(csm.SM())` to here
var (
   err error
   esm = ...
)